Group the Joins and Relations sections in a dedicated chapter (#8435)
qgis/QGIS-Documentation: docs/user_manual/working_with_vector/joins_relations.rst

.. index:: Joins, Relations
.. _sec_joins_relations:

*******************************************
 Connecting and Editing Data Across Layers
*******************************************

.. only:: html

   .. contents::
      :local:

Ability to connect data from different layers is one of the duties of a GIS software.
Such a connection can be based on the spatial relationship between the features,
or on their shared attributes.
QGIS provides tools to handle any of these associations, such as:

* Processing algorithms that can create a new layer as a result of the connection,
  namely :ref:`qgisjoinattributesbylocation`, :ref:`qgisjoinbynearest`,
  :ref:`qgisjoinattributestable`, ...
* SQL queries to create a new layer from the :ref:`DB Manager <dbmanager>`
  or as a :ref:`virtual layer <vector_virtual_layers>`
* :ref:`Joins properties <sec_joins>` or :ref:`relations settings <vector_relations>`
  that temporarily extend attributes of features in a given layer,
  with those of features in another layer based on some matching attribute(s).

  Joins and relations are technical concepts borrowed from databases
  to get the most out of data stored in tables by combining their contents.

>   The idea is that features (rows) of different layers (tables) can belong to each other.

Reviewer: I'm not sure `belong` is the correct term. If you think about composition vs assiociation. What about "can be associated to each other" ?

  The links between the features can be of one-to-one type (joins) or one/many to many (relations).

.. index:: Joins, Foreign key
.. _`sec_joins`:

Joining features between two layers
====================================

**Joins** allow you to associate features of the current layer
to features from another loaded vector layer.

> Whether they are spatially enabled and their geometry type does not matter.

Reviewer: ```suggestion
Whether they are spatially enabled and the type of geometry do not matter.
```

> The join is based on an attribute that is shared by the layers, in a one-to-one relationship.

Reviewer: it's not always one-to-one I believe. One-to-many is also possible.

Reviewer 2: As I understand it, QGIS doesn't have a direct built-in one-to-many join operation for attribute tables. To achieve one-to-many relationship-like results, users need to use SQL queries or spatial databases. Therefore, there is no single workflow in QGIS for one-to-many joins. Consequently, I don't think one-to-many joins can be adequately described without external guidelines (such as PostgreSQL/PostGIS), and I'm not sure if that's something we want to include in QGIS documentation. @DelazJ, you may have more insight into the rules for external guidelines in QGIS docs.

Anyway, while I'm not an expert on spatial databases, I do know that Joins in QGIS allow relationships where features in one layer can be associated with features in another layer based on shared attributes. For instance, you could join a point layer representing cities to a table containing information about each city. Each city point can be associated with one record in the attribute table. Perhaps this is considered a one-to-many join, even if it's not a true one-to-many relationship in a relational database sense. In other words, is it possible that the terms 'one-to-one' and 'one-to-many' are misunderstood?

Author: @selmaVH1  Thanks for the feedback and confirming that joins in layer properties only allow a feature to be associated to one other record.
Speaking of one-to-many relation, this is also possible in QGIS through the project relations but this is already addressed in the other part of the PR.

  * |joinNotEditable|: the join layer is not configured to be editable.
    If you want to be able to edit join features from the target attribute table,
    then you have to check the option |checkbox| :guilabel:`Editable join layer`.
  * |joinedLayerNotEditable|: the join layer is well configured to be editable,
    but its current status is read only.
  * |joinHasNotUpsertOnEdit|: the join layer is editable,
    but synchronization mechanisms are not activated.
    If you want to automatically add a feature in the join layer
    when a feature is created in the target layer,
    then you have to check the option |checkbox| :guilabel:`Upsert on edit`.
    Symmetrically, the option |checkbox| :guilabel:`Delete cascade` may be activated
    if you  want to automatically delete join features.
* |unchecked| :guilabel:`Joined fields`: instead of adding all the fields from the joined layer,
  you can specify a subset.
* |unchecked| :guilabel:`Custom field name prefix` for joined fields,
  in order to avoid name collision

QGIS currently has support for joining non-spatial table formats supported by GDAL
(e.g., CSV, DBF and Excel), delimited text and the PostgreSQL providers.

> .. is the above still true? No more supported formats (oracle, mssql, ...)???

Reviewer: I would have thought that this is available for any data source.

.. index:: Relations, Foreign key
.. _vector_relations:

Setting relations between multiple layers
=========================================

> Unlike joins that define a one-to-one link between features across two layers,

Reviewer: I don't think this is the difference (as stated above, joins are not strictly 1-to-1). To me the main difference is that relation offers a mean to navigate through the relations, in other words list and discover features which are associated to each other.

relations help you build interconnections between multiple features across two or more layers.
As such, relations are project level settings
and are set in :menuselection:`Project --> Properties -->` |relations| :guilabel:`Relations` tab.
From there, you can:

* |symbologyAdd| :guilabel:`Add relation` whose type can be:

  * :ref:`one to many relation <one_to_many_relation>`
  * :ref:`many to many relation <many_to_many_relation>`
  * :ref:`polymorphic relation <polymorphic_relation>` that you can add or edit
    with the dedicated tools in the action drop-down menu.

>   .. note:: There is no simple way yet to edit a non-polymorphic relation once it has been created.

Reviewer: only the name can be changed by double-clik

>   .. note:: There is no simple way yet to edit a non-polymorphic relation once it has been created.
>    To modify a relation you will have to remove and recreate it from scratch.

Author: ```suggestion
  .. note:: There is no simple way yet to edit a non-polymorphic relation once it has been created.
   Only the name can be edited with a double-click.
   For any other parameters of such a relation you will have to remove and recreate it.
```

         DEFERRABLE INITIALLY DEFERRED;

Then in QGIS, you should set up two :ref:`one-to-many relations <one_to_many_relation>`
as explained above:

* a relation between ``airlines`` table and the pivot table;
* and a second one between ``airports`` table and the pivot table.

An easier way to do it (only for PostgreSQL) is using the :guilabel:`Discover Relations`
in :menuselection:`Project --> Properties --> Relations`.
QGIS will automatically read all relations in your database
and you only have to select the two you need.
Remember to load the three tables in the QGIS project first.

.. _figure_setup_relations:

.. figure:: img/airports_airlines_relation.png
   :align: center

   Relations and autodiscover

In case you want to remove an ``airport`` or an ``airline``,
QGIS won't remove the associated record(s) in ``airports_airlines`` table.
This task will be made by the database if we specify the right *constraints*
in the pivot table creation as in the current example.

.. note:: **Combining N-M relation with automatic transaction group**

  You should enable the transaction mode in :menuselection:`Project Properties --> Data Sources -->`
  when working on such context.

>   QGIS should be able to add or update row(s) in all tables (airlines, airports and the pivot tables).

Reviewer: shall we suggest here to use UUIDs instead of increment?

Reviewer 2: @3nids I assume this refers to PostgreSQL, not QGIS?
